chapters/classes.tex

```diff
@@ -217,17 +217,27 @@ \subsubsection{Prefix Rules}\label{prefix-rules}
 
 A \firstuse{prefix} is property of an element of a class definition which can be present or not be present, e.g., \lstinline!final!, \lstinline!public!, \lstinline!flow!.
 
-Variables declared with the \lstinline!flow! or the \lstinline!stream! type prefix shall be a subtype of \lstinline!Real!.
-
 Type prefixes (that is, \lstinline!flow!, \lstinline!stream!, \lstinline!discrete!, \lstinline!parameter!, \lstinline!constant!, \lstinline!input!, \lstinline!output!) shall only be applied for type,
```

Comment:
I have tried to add a clarification that not all of those combinations are legal.

Reply:
> I looked through the code and the test cases I have because I remember implementing the support the way it is described in the paragraph on line 230 (master line numbering). If I change that to not allow flows and streams to be records (as required by the suggested lines 228 and 229) then I get a few test failures. But all of them are the tests that we wrote to test the implementation of this paragraph. So I could not find any real life example. However, how can we be sure that there are no user models that would use that feature?

My logic was that for `flow `the specification previously said:

> Variables declared with the flow or the stream type prefix shall be a subtype of Real

However, we could change it so that 'flow' is applied to elements of structured components, except for operator record and allow that. That would be backward compatible even for those tests.

For `stream `there's the additional requirement that it has to be a scalar Real for the rest of the stream-handling logic; having an exception for a record with a scalar Real just seems contrived - and I don't think it has been implemented.

Reply:
Sorry, last time I had to rush into a meeting and had time to summarize the problem but no time to offer a suggestion. I think with the requirements that connection sets are build up from primitive variables (as defined in Generation of Connect Equation), it is still valid to have structured components declared as flow and streams, as long as the primitive variables are subtype of Real or operator records for flows and subtype of Real for streams. And the additional requirement on flows when there ares streams present. At least, this, I believe, would be enough to fit with other requirements of the specification without contradictions or making the change backward-incompatible. I don't know if it would make sense from engineering/physical point of view to have records of streams. But it does from programmatic point of view.

So, I think, instead of the changes to classes.tex that have been applied so far in this PR, we need to remove the paragraph:
> Variables declared with the flow or the stream type prefix shall be a subtype of Real.

and insert after the paragraph that explains application of prefixes to structured components (line 230 of master):
> The primitive elements with prefix \lstinline!stream! shall be a subtype of Real.
> The primitive elements with prefix \lstinline!flow! shall be a subtype of Real or an operator record component.

Then we should probably make the definition of the primitive elements in Generation of Connect Equation:
> The primitive elements are of simple types or of types defined as operator record (i.e. a component of an operator record type is not split into sub-components).

into a proper definition, so that we can reference it in the suggested text above. The change to streams.tex as suggested in this PR should also be applied. And the paragraph on line 230 should be split into one sentence per line.

Reply:
I will change along those lines for flow; especially as you can have a record with multiple elements.
But not for stream as there are so many formulations that would need to be changed - and since there can anyway only be one scalar stream in a connector, it would only allow a record with one scalar Real element.

Reply:
I have now done that, is it ok?